### PR TITLE
Add rats to Drylake environment

### PR DESCRIPTION
**Pull request by imgntn**
This PR adds a script that creates rats that run down the alley in Drylake.  

To test, go to Drylake and run https://rawgit.com/imgntn/hifi/ratsteer/examples/drylake/ratCreator.js in the alley.

### CODE REVIEW

File: examples/drylake/steer.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    var m = scale(d, 0, ARRIVAL_DISTANCE, 0, MAX_SPEED);
+  } else {
+    desired = Vec3.multiply(MAX_SPEED, desired);
+
+  }
+  var steer = Vec3.subtract(desired, velocity);
+  var steerVector = new V3(desired.x, 0, desired.z);
+  steer = steerVector.limit(MAX_FORCE)
+
+  return steer
```
Comment (ZappoMan) [inline examples/drylake/steer.js]: missing `;`

File: examples/drylake/steer.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  desired = Vec3.normalize(desired);
+  if (d < ARRIVAL_DISTANCE) {
+    var m = scale(d, 0, ARRIVAL_DISTANCE, 0, MAX_SPEED);
+  } else {
+    desired = Vec3.multiply(MAX_SPEED, desired);
+
+  }
+  var steer = Vec3.subtract(desired, velocity);
+  var steerVector = new V3(desired.x, 0, desired.z);
+  steer = steerVector.limit(MAX_FORCE)
```
Comment (ZappoMan) [inline examples/drylake/steer.js]: missing `;`

File: examples/drylake/steer.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
+function flee(thisEntity, target) {
```
Comment (ZappoMan) [inline examples/drylake/steer.js]: missing header - formatting not consistent with coding standard (4 space indents)

File: examples/drylake/ratCreator.js [hunk 4]
```diff
@@ -0,0 +1,493 @@
+Script.include('steer.js');
```
Comment (ZappoMan) [inline examples/drylake/ratCreator.js]: missing header

File: examples/drylake/createAvatarDetector.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
+var avatarDetector = null;
```
Comment (ZappoMan) [inline examples/drylake/createAvatarDetector.js]: missing header

File: examples/drylake/createAvatarDetector.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        }
+    }
+   avatarDetector= Entities.addEntity(detectorProperties);
+};
+var updateAvatarDetector = function() {
+    //  print('updating detector position' + JSON.stringify(MyAvatar.position))
+
+    Entities.editEntity(avatarDetector, {
+        position: MyAvatar.position
+    })
```
Comment (ZappoMan) [inline examples/drylake/createAvatarDetector.js]: missing ';'

File: examples/drylake/ratCreator.js [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+    },
+    visible: false,
+    position: RAT_NEST_LOCATION
+};
+
+var target = Entities.addEntity(targetProperties);
+
+function addRat() {
+    var rat = Entities.addEntity(modelRatProperties);
+    return rat
```
Comment (ZappoMan) [inline examples/drylake/ratCreator.js]: missing `;`

File: examples/drylake/ratCreator.js [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+
+function updateTweens() {
+    TWEEN.update();
+}
+
+function createRatSoundInjector() {
+    var audioOptions = {
+        volume: 0.05,
+        loop: false
+    }
```
Comment (ZappoMan) [inline examples/drylake/ratCreator.js]: missing `;`

File: examples/drylake/ratCreator.js [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+
+    var injector = Audio.playSound(ratRunningSound, audioOptions);
+
+    return injector
```
Comment (ZappoMan) [inline examples/drylake/ratCreator.js]: missing `;`

File: examples/drylake/ratCreator.js [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+
+        //see if there are avatars to run from
+        var avatarFlightVectors = steer.fleeAllAvatars(rat);
+        var averageAvatarFlight;
+        var i;
+        for (i = 0; i < avatarFlightVectors.length; i++) {
+            if (i === 0) {
+                averageAvatarFlight = avatarFlightVectors[0];
+            } else {
+                averageAvatarFlight = Vec3.sum(avatarFlightVectors[i - 1], avatarFlightVectors[i])
```
Comment (ZappoMan) [inline examples/drylake/ratCreator.js]: missing `;`

File: examples/drylake/ratCreator.js [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+
+        //see if there are avoiders to flee
+        var avoidBlockVectors = steer.fleeAvoiderBlocks(rat);
+        var averageAvoiderFlight;
+        var j;
+        for (j = 0; j < avoidBlockVectors.length; j++) {
+            if (j === 0) {
+                averageAvoiderFlight = avoidBlockVectors[0];
+            } else {
+                averageAvoiderFlight = Vec3.sum(avoidBlockVectors[j - 1], avoidBlockVectors[j])
```
Comment (ZappoMan) [inline examples/drylake/ratCreator.js]: missing `;`

File: examples/drylake/ratCreator.js [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+        averageVector = Vec3.multiply(averageVector, 1 / divisorCount);
+
+        var thisRatProps = Entities.getEntityProperties(rat, ["position", "rotation"]);
+        var ratPosition = thisRatProps.position;
+        var ratToNest = Vec3.subtract(RAT_NEST_LOCATION, ratPosition);
+        var ratRotation = Quat.rotationBetween(Vec3.UNIT_Z, ratToNest);
+        var eulerAngle = Quat.safeEulerAngles(ratRotation);
+        eulerAngle.x = 0;
+        eulerAngle.z = 0;
+        var constrainedRotation = Quat.fromVec3Degrees(eulerAngle)
```
Comment (ZappoMan) [inline examples/drylake/ratCreator.js]: missing `;`

File: examples/drylake/ratCreator.js [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+
+        Entities.editEntity(rat, {
+            velocity: averageVector,
+            rotation: constrainedRotation,
+        })
```
Comment (ZappoMan) [inline examples/drylake/ratCreator.js]: missing `;`

File: examples/drylake/ratCreator.js [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+        //have to make a 'meta' rat object to keep track of rats for updating sound injector locations.  parenting sounds would make this easy.
+        var metaRat = getMetaRatByRat(rat);
+        if (metaRat !== undefined) {
+            if (metaRat.injector !== undefined) {
+                if (metaRat.injector.isPlaying === true) {
+                    //  print('update injector position')
+                    metaRat.injector.options = {
+                        loop: true,
+                        position: ratPosition
+                    }
```
Comment (ZappoMan) [inline examples/drylake/ratCreator.js]: missing `;`

File: examples/drylake/ratSteer.js [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+//  Created by James B. Pollack @imgntn on 12/7/2015
+//  Copyright 2015 High Fidelity, Inc.
+//
+//  This is an example of steering behaviors that can be applied entities.
+//
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+function flee(thisEntity, target) {
```
Comment (ZappoMan) [inline examples/drylake/ratSteer.js]: indents in this file are wrong... it should be 4 spaces

File: examples/drylake/ratSteer.js [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+  return min2 + (max2 - min2) * ((value - min1) / (max1 - min1));
+}
+
+loadSteer = function() {
+  return {
+    flee: flee,
+    fleeAllAvatars: fleeAllAvatars,
+    fleeAvoiderBlocks: fleeAvoiderBlocks,
+    arrive: arrive
+  }
```
Comment (ZappoMan) [inline examples/drylake/ratSteer.js]: missing `;`

File: examples/drylake/ratSteer.js [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+
+      var desired = Vec3.subtract(location, entityProps.position);
+      var d = Vec3.length(desired);
+      desired = Vec3.normalize(desired);
+      desired = Vec3.multiply(MAX_SPEED, desired);
+
+      if (d < 5) {
+        var steer = Vec3.subtract(desired, velocity);
+        var steerVector = new V3(desired.x, 0, desired.z);
+        steer = steerVector.limit(MAX_FORCE)
```
Comment (ZappoMan) [inline examples/drylake/ratSteer.js]: missing `;`

File: examples/drylake/ratSteer.js [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+      //print('found an avatar to flee')
+
+      var MAX_SPEED = 8;
+      var MAX_FORCE = 8;
+
+      var desired = Vec3.subtract(location, entityProps.position);
+      var d = Vec3.length(desired);
+      desired = Vec3.normalize(desired);
+      desired = Vec3.multiply(MAX_SPEED, desired);
+      if (d < 3) {
```
Comment (ZappoMan) [inline examples/drylake/ratSteer.js]: magic number?

File: examples/drylake/ratCreator.js [hunk 19]
```diff
@@ -0,0 +1,474 @@
...
+
+var ratSpawnerInterval;
+
+if (USE_CONSTANT_SPAWNER === true) {
+    ratSpawnerInterval = Script.setInterval(function() {
+        var rat = addRat();
+        playRatRunningAnimation(rat);
+        rats.push(rat);
+        ratCount++;
+        if (ratCount % 6 === 0) {
```
Comment (ZappoMan) [inline examples/drylake/ratCreator.js]: magic number

File: examples/drylake/ratCreator.js [hunk 20]
```diff
@@ -0,0 +1,474 @@
...
+                rat: rat,
+                injector: createRatSoundInjector()
+            }
+            metaRats.push(metaRat);
+
+            Script.setTimeout(function() {
+                //if we have too many injectors hanging around there are problems
+                metaRat.injector.stop();
+                delete metaRat.injector;
+            }, RAT_SPAWN_RATE * 3);
```
Comment (ZappoMan) [inline examples/drylake/ratCreator.js]: magic number

**Comment by hifi-gustavo** [comment]
Build console output for commit: 99bc47867ac769b1d27c7028096c5bac2c9058f3 available at: https://gist.github.com/d8a56a1a9128c0693d50<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/High+Fidelity-PR-6503-win64.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6503.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a31ba089c1623e897f74e18126614ce4009f8a4e available at: https://gist.github.com/d8a56a1a9128c0693d50<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/High+Fidelity-PR-6503-win64.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6503.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 35da54fc6c910964f48a406922db2d0d4a7ae51d available at: https://gist.github.com/d8a56a1a9128c0693d50<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/High+Fidelity-PR-6503-win64.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6503.dmg

**Pull request merged**